Cargo: Use json output to get better error messages

JelteF opened the pull request:
This modifies the cargo maker to use its json output. This makes it much easier
to parse than the human friendly version. 

It currently uses python to parse the JSON in a safe way. This obviously
requires +python to be available, which seems like a reasonable tradeoff.

I'm planning to add support for the length field in a future pull request/commit.

blueyed commented on autoload/neomake/makers/cargo.vim:
```diff
@@ -2,8 +2,33 @@
+    if l:text[0] == '{'
+        echom len(l:text)
+        python import json
+        let l:data = pyeval("json.loads(vim.eval('l:text'))")['message']
```
Vim's `json_decode` cannot be used here?

blueyed commented on autoload/neomake/makers/cargo.vim:
```diff
@@ -2,8 +2,33 @@
+        python import json
+        let l:data = pyeval("json.loads(vim.eval('l:text'))")['message']
+        let l:code = l:data['code']
+        echom type(l:code)
```
leftover.

JelteF commented:
Good one, I didn't know about `json_decode`. Changed it to use that.

JelteF commented:
There seems to be a bug when there are no errors, so don't merge this yet.

blueyed commented:
Please consider adding tests for this.
See the latest `ft_*` ones.

blueyed commented:
@JelteF 
Do you like to add some tests?
Otherwise, please provide some example output, so that somebody else could add them.
Since this is the first maker using JSON, it should be covered.

JelteF commented:
I'll write some tests. 

Also, what do you think about removing the regular rustc maker and replacing it with the cargo one. 
The rustc one doesn't work for any of my projects that use dependencies, which is why I have it disabled. Cargo is the advised build tool for rust projects anyway and it's very weird to not have it installed.

blueyed commented:
> Also, what do you think about removing the regular rustc maker and replacing it with the cargo one.
The rustc one doesn't work for any of my projects that use dependencies, which is why I have it disabled. Cargo is the advised build tool for rust projects anyway and it's very weird to not have it installed.

Makes sense.
I'd say to keep `rustc`, but to not enable it by default.

JelteF commented:
I added some simple tests and moved the maker. However, with the current tests as examples I couldn't get the result of passing the errorformat over the generated error string.

blueyed commented on tests/ft_cargo.vader:
```diff
@@ -0,0 +1,21 @@
+Include: include/setup.vader
```
Not sure about the `ft_` prefix here, since it's not a filetype maker after all.
Rename it to `cargo.vader`?

blueyed commented on tests/ft_cargo.vader:
```diff
@@ -0,0 +1,21 @@
+  let &errorformat = neomake#makers#cargo#cargo().errorformat
+  let error = neomake#makers#cargo#CargoParseJSON(json)
+
+  AssertEqual error, 0
```
It is `""` really, isn't it?

blueyed commented on autoload/neomake/makers/ft/rust.vim:
```diff
@@ -22,3 +22,33 @@ function! neomake#makers#ft#rust#rustc() abort
+        \ 'args': ['test', '--no-run', '--message-format=json', '--quiet'],
+        \ 'errorformat':
+            \ '[%t%n] "%f" %l:%v %m',
+        \ 'mapexpr': 'neomake#makers#cargo#CargoParseJSON(v:val)',
```
It should have `'append_file': 0` probably?!

blueyed commented:
There is one error:
> Vim(function):E746: Function name does not match script file name: neomake#makers#cargo#CargoParseJSON

Also `ft_rust.vader` should be included in `neomake.vader`.

blueyed commented:
> ( 81/101) [EXECUTE] (X) Vim(let):E117: Unknown function: json_decode

Can we maybe just fallback on rustc for older Vim?

blueyed commented:
The best would be to fallback to the old cargo maker in this case, but I am happy to wait for a complaint first?!

blueyed commented on autoload/neomake/makers/ft/rust.vim:
```diff
@@ -35,7 +35,13 @@ endfunction
+            python import json
+            let l:decoded = pyeval("json.loads(vim.eval('l:text'))")
+        endif
+        let l:data = get(l:decoded, 'message', {})
```
Nice one! :)
Forgot about it already.

JelteF commented:
Weird, the tests still fail. They don't seem to have python support, I've never used regular vim without it.

blueyed commented on Dockerfile.tests:
```diff
@@ -13,8 +13,8 @@ RUN apk --update add make openssl python \
-        -tag v7.3.429 -name vim73 -build \
-        -tag v7.4.052 -name vim74-trusty -build \
+        -tag v7.3.429 -py -name vim73 -build \
+        -tag v7.4.052 -py -name vim74-trusty -build \
```
That needs to be rebuilt/pushed.
I think it is good to have of without though?!

JelteF replied:
I'm quite sure vim on trusty has python support and for vim 73 I don't know.

But maybe it's better to just ignore these tests for old versions of vim. I'm not sure how to do that though. If you could change that it would be great.

blueyed replied:
https://github.com/neomake/neomake/pull/942#issuecomment-273927592 and the next.

JelteF replied:
I could, but that still won't fix the tests for them, because those require parsing of json strings.

blueyed replied:
The tests should be skipped for this then.
There are skip examples in other tests.

blueyed commented on autoload/neomake/makers/ft/rust.vim:
```diff
@@ -42,10 +43,20 @@ function! neomake#makers#ft#rust#CargoParseJSON(val) abort
-        let l:code = get(l:data, 'code', v:null)
-        if type(l:code) == type({})
-            let l:code = l:code['code']
+        let l:data = get(l:decoded, 'message', v:null)
```
`v:null` won't work on older versions, does it?

JelteF replied:
Could very well be, how should it be done then?

blueyed commented on autoload/neomake/makers/ft/rust.vim:
```diff
@@ -42,10 +43,20 @@ function! neomake#makers#ft#rust#CargoParseJSON(val) abort
-            let l:code = l:code['code']
+        let l:data = get(l:decoded, 'message', v:null)
+        if type(l:data) == type({})
+            echom 'hoooi'
```
You like your `echom`s on Travis apparently.. ;)

blueyed commented:
@JelteF 
There is a test failure with neovim (and others):

> ( 82/102) [EXECUTE] (X) '[E0308] "src/lib.rs" 19:46 mismatched types' should be equal to '[E0308] "src/lib.rs" 19:46 mismatched types: expected str, found struct `proc_macro::TokenStream`'

https://travis-ci.org/neomake/neomake/jobs/194237027#L1427

JelteF commented:
Probably a regression in one of my latest changes. Good thing I added those
tests ;). I'll check it out tomorrow and skip the tests for bad vims as
well.

On Mon, 23 Jan 2017, 23:11 Daniel Hahler, <notifications@github.com> wrote:

> @JelteF <https://github.com/JelteF>
> There is a test failure with neovim (and others):
>
> ( 82/102) [EXECUTE] (X) '[E0308] "src/lib.rs" 19:46 mismatched types'
> should be equal to '[E0308] "src/lib.rs" 19:46 mismatched types: expected
> str, found struct proc_macro::TokenStream'
>
> https://travis-ci.org/neomake/neomake/jobs/194237027#L1427
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neomake/neomake/pull/942#issuecomment-274633902>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABG8JvVU3VlelOW5Ej5vZ3o8tsiUymSoks5rVSVwgaJpZM4LmTf0>
> .
>

JelteF commented:
I'm not sure what is causing the current error. @blueyed do you have any idea?

blueyed commented:
Pushed a fixup, let's see.

blueyed commented:
For reference:
```
( 81/103) [EXECUTE] cargo info message
    ( 81/103) [EXECUTE] (X) Vim(let):Traceback (most recent call last):
      > function neomake#makers#ft#rust#CargoParseJSON, line 7
```

blueyed commented:
For the tests: the skipping should be logged:

```diff
 tests/ft_rust.vader | 118 +++++++++++++++++++++++++++-------------------------
 1 file changed, 62 insertions(+), 56 deletions(-)

diff --git i/tests/ft_rust.vader w/tests/ft_rust.vader
index 7660788..53ccc13 100644
--- i/tests/ft_rust.vader
+++ w/tests/ft_rust.vader
@@ -1,90 +1,96 @@
 Include: include/setup.vader
 
 Execute (cargo info message):
+  if !(exists('*json_decode') || has('python'))
+    Log "SKIP: no JSON support."
+    finish
+  endif
+
   let json = '{"features":["default"],"filenames":["/home/jelte/fun/derive_more/target/debug/deps/libunicode_xid-2c56e0ed90f93244.rlib"],"package_id":"unicode-xid 0.0.3 (registry+https://github.com/rust-lang/crates.io-index)","profile":{"debug_assertions":true,"debuginfo":true,"opt_level":"0","test":false},"reason":"compiler-artifact","target":{"kind":["lib"],"name":"unicode-xid","src_path":"/home/jelte/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-xid-0.0.3/src/lib.rs"}}'
 
-  if exists('*json_decode') || has('python')
+  Save &errorformat
+  let &errorformat = neomake#makers#ft#rust#cargo().errorformat
+  let error = neomake#makers#ft#rust#CargoParseJSON(json)
 
-    Save &errorformat
-    let &errorformat = neomake#makers#ft#rust#cargo().errorformat
-    let error = neomake#makers#ft#rust#CargoParseJSON(json)
+  AssertEqual 0, error
 
-    AssertEqual 0, error
+Execute (cargo error without span):
+  if !(exists('*json_decode') || has('python'))
+    Log "SKIP: no JSON support."
+    finish
   endif
 
-Execute (cargo error without span):
   let json = '{"message":{"children":[],"code":null,"level":"error","message":"aborting due to 18 previous errors","rendered":null,"spans":[]},"package_id":"derive_more 0.4.0 (path+file:///home/jelte/fun/derive_more)","reason":"compiler-message","target":{"kind":["proc-macro"],"name":"derive_more","src_path":"/home/jelte/fun/derive_more/src/lib.rs"}}'
 
-  if exists('*json_decode') || has('python')
+  Save &errorformat
+  let &errorformat = neomake#makers#ft#rust#cargo().errorformat
+  let error = neomake#makers#ft#rust#CargoParseJSON(json)
 
-    Save &errorformat
-    let &errorformat = neomake#makers#ft#rust#cargo().errorformat
-    let error = neomake#makers#ft#rust#CargoParseJSON(json)
-
-    AssertEqual 0, error
-  endif
+  AssertEqual 0, error
 
 
 Execute (cargo error message):
+  if !(exists('*json_decode') || has('python'))
+    Log "SKIP: no JSON support."
+    finish
+  endif
+
   let json = '{"message":{"children":[{"children":[],"code":null,"level":"note","message":"expected type `&str`","rendered":null,"spans":[]},{"children":[],"code":null,"level":"note","message":"   found type `&proc_macro::TokenStream`","rendered":null,"spans":[]}],"code":{"code":"E0308","explanation":"\nThis error occurs when the compiler was unable to infer the concrete type of a\nvariable. It can occur for several cases, the most common of which is a\nmismatch in the expected type that the compiler inferred for a variables\ninitializing expression, and the actual type explicitly assigned to the\nvariable.\n\nFor example:\n\n```compile_fail,E0308\nlet x: i32 = \"I am not a number!\";\n//     ~~~   ~~~~~~~~~~~~~~~~~~~~\n//      |             |\n//      |    initializing expression;\n//      |    compiler infers type `&str`\n//      |\n//    type `i32` assigned to variable `x`\n```\n"},"level":"error","message":"mismatched types","rendered":null,"spans":[{"byte_end":437,"byte_start":435,"column_end":48,"column_start":46,"expansion":{"def_site_span":{"byte_end":537,"byte_start":181,"column_end":3,"column_start":1,"expansion":null,"file_name":"src/lib.rs","is_primary":false,"label":null,"line_end":23,"line_start":14,"suggested_replacement":null,"text":[{"highlight_end":28,"highlight_start":1,"text":"macro_rules! create_derive("},{"highlight_end":55,"highlight_start":1,"text":"    ($mod_:ident, $trait_:ident, $fn_name: ident) => {"},{"highlight_end":38,"highlight_start":1,"text":"        #[proc_macro_derive($trait_)]"},{"highlight_end":61,"highlight_start":1,"text":"        pub fn $fn_name(input: TokenStream) -> TokenStream {"},{"highlight_end":27,"highlight_start":1,"text":"            let s = input;"},{"highlight_end":59,"highlight_start":1,"text":"            let ast = syn::parse_macro_input(&s).unwrap();"},{"highlight_end":70,"highlight_start":1,"text":"            $mod_::expand(&ast, stringify!($trait_)).parse().unwrap()"},{"highlight_end":10,"highlight_start":1,"text":"        }"},{"highlight_end":6,"highlight_start":1,"text":"    }"},{"highlight_end":3,"highlight_start":1,"text":");"}]},"macro_decl_name":"create_derive!","span":{"byte_end":579,"byte_start":539,"column_end":41,"column_start":1,"expansion":null,"file_name":"src/lib.rs","is_primary":false,"label":null,"line_end":25,"line_start":25,"suggested_replacement":null,"text":[{"highlight_end":41,"highlight_start":1,"text":"create_derive!(from, From, from_derive);"}]}},"file_name":"src/lib.rs","is_primary":true,"label":"expected str, found struct `proc_macro::TokenStream`","line_end":19,"line_start":19,"suggested_replacement":null,"text":[{"highlight_end":48,"highlight_start":46,"text":"            let ast = syn::parse_macro_input(&s).unwrap();"}]}]},"package_id":"derive_more 0.4.0 (path+file:///home/jelte/fun/derive_more)","reason":"compiler-message","target":{"kind":["proc-macro"],"name":"derive_more","src_path":"/home/jelte/fun/derive_more/src/lib.rs"}}'
 
-  if exists('*json_decode') || has('python')
+  Save &errorformat
+  let &errorformat = neomake#makers#ft#rust#cargo().errorformat
 
-    Save &errorformat
-    let &errorformat = neomake#makers#ft#rust#cargo().errorformat
+  let error = neomake#makers#ft#rust#CargoParseJSON(json)
 
-    let error = neomake#makers#ft#rust#CargoParseJSON(json)
+  AssertEqual '[E0308] "src/lib.rs" 19:46 2 mismatched types: expected str, found struct `proc_macro::TokenStream`', error
 
-    AssertEqual '[E0308] "src/lib.rs" 19:46 2 mismatched types: expected str, found struct `proc_macro::TokenStream`', error
+  cgetexpr error
 
-    cgetexpr error
+  let result = getqflist()[0]
+  unlet result.bufnr
+  call neomake#makers#ft#rust#CargoPostProcess(result)
 
-    let result = getqflist()[0]
-    unlet result.bufnr
-    call neomake#makers#ft#rust#CargoPostProcess(result)
+  AssertEqual {
+    \ 'type': 'E',
+    \ 'nr': 308,
+    \ 'lnum': 19,
+    \ 'col': 46,
+    \ 'length': 2,
+    \ 'valid': 1,
+    \ 'vcol': 1,
+    \ 'pattern': '',
+    \ 'text': 'mismatched types: expected str, found struct `proc_macro::TokenStream`'}, result
 
-    AssertEqual {
-      \ 'type': 'E',
-      \ 'nr': 308,
-      \ 'lnum': 19,
-      \ 'col': 46,
-      \ 'length': 2,
-      \ 'valid': 1,
-      \ 'vcol': 1,
-      \ 'pattern': '',
-      \ 'text': 'mismatched types: expected str, found struct `proc_macro::TokenStream`'}, result
+Execute (cargo warning message):
+  if !(exists('*json_decode') || has('python'))
+    Log "SKIP: no JSON support."
+    finish
   endif
 
-Execute (cargo warning message):
   let json = '{"message":{"children":[],"code":null,"level":"warning","message":"unused import: `Variant`, #[warn(unused_imports)] on by default","rendered":null,"spans":[{"byte_end":8504,"byte_start":8497,"column_end":31,"column_start":24,"expansion":null,"file_name":"src/add_assign_like.rs","is_primary":true,"label":null,"line_end":2,"line_start":2,"suggested_replacement":null,"text":[{"highlight_end":31,"highlight_start":24,"text":"use syn::{Body, Ident, Variant, VariantData, MacroInput};"}]}]},"package_id":"derive_more 0.4.0 (path+file:///home/jelte/fun/derive_more)","reason":"compiler-message","target":{"kind":["proc-macro"],"name":"derive_more","src_path":"/home/jelte/fun/derive_more/src/lib.rs"}}'
 
-  if exists('*json_decode') || has('python')
-
-    Save &errorformat
-    let &errorformat = neomake#makers#ft#rust#cargo().errorformat
+  Save &errorformat
+  let &errorformat = neomake#makers#ft#rust#cargo().errorformat
 
-    let error = neomake#makers#ft#rust#CargoParseJSON(json)
+  let error = neomake#makers#ft#rust#CargoParseJSON(json)
 
 
-    AssertEqual '[W] "src/add_assign_like.rs" 2:24 7 unused import: `Variant`, #[warn(unused_imports)] on by default', error
+  AssertEqual '[W] "src/add_assign_like.rs" 2:24 7 unused import: `Variant`, #[warn(unused_imports)] on by default', error
 
-    cgetexpr error
+  cgetexpr error
 
-    let result = getqflist()[0]
-    unlet result.bufnr
+  let result = getqflist()[0]
+  unlet result.bufnr
 
-    call neomake#makers#ft#rust#CargoPostProcess(result)
-
-    AssertEqual {
-      \ 'type': 'W',
-      \ 'nr': -1,
-      \ 'lnum': 2,
-      \ 'col': 24,
-      \ 'length': 7,
-      \ 'valid': 1,
-      \ 'vcol': 1,
-      \ 'pattern': '',
-      \ 'text': 'unused import: `Variant`, #[warn(unused_imports)] on by default'}, result
-
-  endif
+  call neomake#makers#ft#rust#CargoPostProcess(result)
 
+  AssertEqual {
+    \ 'type': 'W',
+    \ 'nr': -1,
+    \ 'lnum': 2,
+    \ 'col': 24,
+    \ 'length': 7,
+    \ 'valid': 1,
+    \ 'vcol': 1,
+    \ 'pattern': '',
+    \ 'text': 'unused import: `Variant`, #[warn(unused_imports)] on by default'}, result
```

I could push it, but feel free to apply it directly.

JelteF commented:
LGTM, feel free to push it

blueyed commented:
Now it is:
> ( 81/103) [EXECUTE] (X) Vim(let):E121: Undefined variable: v:null
>    > function neomake#makers#ft#rust#CargoParseJSON, line 9

Will push another fixup.

blueyed commented:
Next round.
Seems like the Vim on Travis did not like the `pyeval` (since https://github.com/neomake/neomake/pull/942/commits/3f1205791e2c316df69b965c995a7544cc132e53 fixed it).
(fixed in 89c5606 just now)

HTH, I'm out for today.

blueyed commented:
Tried to see the error, but it's not in `:mess`?!  (https://travis-ci.org/neomake/neomake/builds/194993254#L1163).

I guess the next step would be running a debugging session on Travis (which only travis-ci.com seems to support) and figure out what is happening.

Could do that tomorrow.

Hope you don't mind the mess in your branch until then.. ;)

blueyed commented:
Fixed the first test, and reached "E859: Failed to convert returned python object to vim value: " because it fails to convert `null`.

I would say it's not worth it..!

blueyed commented:
I am going to activate Python in some of the Docker Vim versions, and then it should be easy to replicate there.  (Travis has not Vim 8.0.0056, but only that patch additionally)

blueyed commented:
Meh.. we had Python enabled builds here already..

```diff
diff --git a/Dockerfile.tests b/Dockerfile.tests
index d20a2d9..b6fe17f 100644
--- a/Dockerfile.tests
+++ b/Dockerfile.tests
@@ -12,8 +12,8 @@ RUN apk --update add make openssl python \
 #  - v7.4.1689 (Ubuntu Xenial, 16.04LTS)
 #  - v8.0.8069 (Updated Vim 8, https://groups.google.com/forum/#!topic/vim_dev/ro_FpeXgc24)
 RUN install_vim \
-        -tag v7.3.429 -name vim73 -build \
-        -tag v7.4.052 -name vim74-trusty -build \
+        -tag v7.3.429 -py -name vim73 -build \
+        -tag v7.4.052 -py -name vim74-trusty -build \
         -tag v7.4.1689 -name vim74-xenial -build \
         -tag v8.0.0069 -name vim8069 -build
```

Sorry, apparently I cannot think straight anymore right now.. :/

I will still rebase this against master.

JelteF commented:
I rebased and added some changes again. But it seems pyeval is broken on the testvim version of vim. Maybe we should just exclude it.

jedahan commented:
Just a note, this fixed all my issue with neomake + rust for me, thank you JelteF

blueyed commented:
Rebased for now.
We should just let old Vims behind probably.

blueyed commented:
This should be rebased on https://github.com/neomake/neomake/pull/1048, which brings `neomake#utils#JSONdecode`.

blueyed commented:
Rebased on https://github.com/neomake/neomake/pull/1048 (awaiting feedback before merging it).

JelteF commented:
I just rebased onto master and (hopefully) fixed all the tests. I think this can now be merged. I think #1050 might indeed be used for the tests for the cargo maker, but I don't want to block this pull request because on #1050 and #1057. It's fine for now and once the new feature is added the tests can easily be changed.

blueyed commented:
The neovim build should be fixed by now (or rather after the next nightly build): https://github.com/neovim/neovim/commit/eb8cbfc8f514c5c84741f782314dc9c256b6d977.

blueyed commented on .travis.yml:
```diff
@@ -24,9 +24,9 @@ env:
-    - ENV=check
     - ENV=vint-diff
     - ENV=vint-errors
+    - ENV=vimlint-errors
```
This seems wrong: check should stay, and vimlint-errors was removed.

blueyed commented on .travis.yml:
```diff
@@ -42,7 +42,7 @@ install:
       # Run full vint against changed files.
       echo "Looking for changed files in $TRAVIS_COMMIT_RANGE."
-      CHANGED_VIM_FILES=($(git diff --name-only "$TRAVIS_COMMIT_RANGE" | grep '\.vim$'))
+      CHANGED_VIM_FILES=($(git diff --name-only --diff-filter=AM "$TRAVIS_COMMIT_RANGE" | grep '\.vim$'))
```
Will cherry-pick this to master: 4e093f7.

blueyed commented on .travis.yml:
```diff
@@ -60,6 +60,12 @@ install:
+      python -V
+      python2 -V
+    fi
+
```
Debugging I guess?  Should be removed.
But might be useful to have/keep in general?!  What do you think?
(I've added it in ac84dc8ac)

JelteF replied:
I think this might indeed be useful

blueyed commented:
> I think #1050 might indeed be used for the tests for the cargo maker, but I don't want to block this pull request because on #1050 and #1057

It is not meant to be used for the tests, but to skip the whole mapexpr: currently you have to use mapexpr to massage the entries for errorformat.  With #1050 you can skip this, and just return/provide the location list entries directly.
But this can be improved later, although it might be me doing this then again.. :/

brycefisher commented:
Super exciting work in PR! Can't wait for it to land!!

JelteF commented:
I fixed the issue with travis and created a new pull request that uses the changes from #1050, see #1123.

JelteF commented:
@blueyed I think this should be fine to merge into master as it works great. That way a pull request for #843 can also be created based on this PR.

Let's continue rewriting the code to use the new API though as it is indeed much nicer. However, I'm running into some bugs hat are not present in this version, see #1123  (probably since the API is new, or my usage is wrong). Working out these kinks will probably take a couple of more weeks as I don't have that much time to work on it and I require feedback from you.

blueyed commented:
I'd rather like to wait for #1123.

blueyed commented:
@JelteF 
What was the reason to make this a file mode maker?

JelteF commented:
After understanding the plugin better the reason wasn't really good. It was
just to have it as the default maker for rust files, because the rustc
maker didn't actually work.

On Sat, Apr 29, 2017, 19:28 Daniel Hahler <notifications@github.com> wrote:

> @JelteF <https://github.com/JelteF>
> What was the reason to make this a file mode maker?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neomake/neomake/pull/942#issuecomment-298178951>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABG8JqYYyz65oJ83ok5gOUQiNEi0WxUlks5r02UYgaJpZM4LmTf0>
> .
>

blueyed commented:
@JelteF 
Thanks for letting me know.  I did not take attention to this myself back then.

I think we should revert this, and rather make it easier to use project makers for filetypes automatically (somehow the reverse of https://github.com/neomake/neomake/pull/790).

I am working on the automake branch (#1167) to make it easier to configure when linting should happen, so that could be solved by that.
